pkg/resourcemonitor/podresourcesscanner.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcemonitor
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	podresourcesapi "k8s.io/kubelet/pkg/apis/podresources/v1"
+	"sigs.k8s.io/node-feature-discovery/pkg/apihelper"
+)
+
+type PodResourcesScanner struct {
+	namespace         string
+	podResourceClient podresourcesapi.PodResourcesListerClient
+	apihelper         apihelper.APIHelpers
+}
+
+func NewPodResourcesScanner(namespace string, podResourceClient podresourcesapi.PodResourcesListerClient, kubeApihelper apihelper.APIHelpers) (ResourcesScanner, error) {
+	resourcemonitorInstance := &PodResourcesScanner{
+		namespace:         namespace,
+		podResourceClient: podResourceClient,
+		apihelper:         kubeApihelper,
+	}
+	if resourcemonitorInstance.namespace != "" {
+		klog.Infof("watching namespace %q", resourcemonitorInstance.namespace)
+	} else {
+		klog.Infof("watching all namespaces")
+	}
+
+	return resourcemonitorInstance, nil
+}
+
+// isWatchable tells if the the given namespace should be watched.
+func (resMon *PodResourcesScanner) isWatchable(podNamespace string, podName string, hasDevice bool) (bool, bool, error) {
+	cli, err := resMon.apihelper.GetClient()
+	if err != nil {
+		return false, false, err
+	}
+	pod, err := resMon.apihelper.GetPod(cli, podNamespace, podName)
+	if err != nil {
+		return false, false, err
+	}
+
+	klog.Infof("podresource: %s", podName)
+	isIntegralGuaranteed := podGuaranteedCPUs(pod)
+
+	if resMon.namespace == "" && (isIntegralGuaranteed || hasDevice) {
+		return true, isIntegralGuaranteed, nil
+	}
+	// TODO:  add an explicit check for guaranteed pods and pods with devices
+	return resMon.namespace == podNamespace && (isIntegralGuaranteed || hasDevice), isIntegralGuaranteed, nil
+}
+
+func podGuaranteedCPUs(pod *v1.Pod) bool {
+	var totalCPU int64
+	var cpuQuantity resource.Quantity
+	for _, container := range pod.Spec.InitContainers {
+
+		var ok bool
+		if cpuQuantity, ok = container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		totalCPU += cpuQuantity.Value()
+		isInitContainerGuaranteed := guaranteedCPUs(pod, &container)
+		if !isInitContainerGuaranteed {
+			return false
```

Comment:
I don't quite understand the logic of this function. We return false if any of the containers requests fractional cpus(?)

Reply:
Yes that is correct, if any container has fractional CPU it means that the pod is not allocated CPUs exclusively and hence we don’t consider that pod for accounting and return false. We had to explicitly handle the case where no CPUs are requested in any of the containers in the pod.

Reply:
Of course, sorry, my bad 😇